### PR TITLE
clustermesh-apiserver deployment: support lifecycle and terminationGr…

### LINKED ISSUE #26900: CFP: clustermesh-apiserver deployment support lifecycle and terminationGracePeriodSeconds
## Cilium Feature Proposal

**Is your feature request related to a problem?**
Yes.
If you enable cilium clustermesh in AWS environment and also use NLB to expose apiserver,  you will encounter some connection issue when terminating clustermesh-apiserver pod. Because NLB will need some time to draining target(apiserver) but kubernetes already sent signal term to the apiserver pod to close process.

**Describe the feature you'd like**
Helm chart templates/clustermesh-apiserver/deployment.yaml support
[1] `lifecycle` in apiserver and etcd container
[2] clustermesh-apiserver pod support `terminationGracePeriodSeconds`

**(Optional) Describe your proposed solution**

In this scenario, we need to set lifecycle to the apiserver and etcd container and also need to increase clustermesh-apiserver pod's terminationGracePeriodSeconds for handle graceful shutdown.

### PR DESCRIPTION
**Pull request by acgs771126**
clustermesh-apiserver deployment: support lifecycle and terminationGracePeriodSeconds

clustermesh-apiserver deployment support lifecycle and terminationGracePeriodSeconds, because we need a while time(by lifecycle.preStop hook) to wait AWS NLB draining target for prevent connection issue when clustermesh-apiserver pod terminating

Please note:  this change I hope can backported to 1.13 version

Fixes: #26900

```release-note
clustermesh-apiserver deployment support lifecycle and terminationGracePeriodSeconds.
```

### CODE REVIEW

**Comment by giorio94** [comment]
Related discussion in https://github.com/cilium/cilium/issues/26900.

File: install/kubernetes/cilium/values.yaml [hunk 1]
```diff
@@ -2801,6 +2804,12 @@ clustermesh:
     # -- Number of replicas run for the clustermesh-apiserver deployment.
     replicas: 1
 
+    # -- lifecycle setting for the apiserver container.
+    lifecycle: {}
+
+    # -- terminationGracePeriodSeconds(default 30s) for the clustermesh-apiserver deployment.
```
Comment (kaworu) [inline install/kubernetes/cilium/values.yaml]: nit: no need to specify the default, the helm doc table has a column for the value and when reading the value file we can read the actual default value.
```suggestion
    # -- terminationGracePeriodSeconds for the clustermesh-apiserver deployment.
```

**Review by kaworu — DISMISSED** [review]
Thanks for the PR @acgs771126!

[CI failed](https://app.travis-ci.com/github/cilium/cilium/builds/264648114#L563) with:

> HINT: to fix this, run 'make -C Documentation update-helm-values'


Also please first report the changes from `install/kubernetes/cilium/values.yaml` to `install/kubernetes/cilium/values.yaml.tmpl` instead and then run:

```console
make -C install/kubernetes cilium/values.yaml
```

File: install/kubernetes/cilium/values.yaml [hunk 2]
```diff
@@ -2721,6 +2721,9 @@ clustermesh:
       # -- Security context to be added to clustermesh-apiserver etcd containers
       securityContext: {}
 
+      # -- lifecycle setting for the etcd container.
+      lifecycle: {}
```
Comment (giorio94) [inline install/kubernetes/cilium/values.yaml]: Please add the the lifecycle configuration also to the kvstoremesh container below.

**Review by giorio94 — CHANGES_REQUESTED** [review]
Thanks @acgs771126 for the PR!

One minor request inline on top of the comments from Alexandre.

File: install/kubernetes/cilium/values.yaml (hunk 1 above)
Comment (acgs771126) [reply install/kubernetes/cilium/values.yaml]: okay, let me update the PR, thanks.

**Review by giorio94 — CHANGES_REQUESTED** [review]
Changes look good to me.

CI is failing with the following errors, which shall be fixed:

```
=========================================================
[1/1] Running on 6f0006779a9ecc757e06931cbe5facee053c0e43
clustermesh-apiserver deployment: support lifecycle and terminationGracePeriodSeconds
=========================================================
ERROR:NO_AUTHOR_SIGN_OFF: Missing Signed-off-by: line by nominal patch author 'andychuang <andy.chuang@shoplineapp.com>'

"[PATCH] clustermesh-apiserver deployment: support lifecycle and" has style problems, please review.

ERROR:CUSTOM: Please avoid long commit subjects (max: 75, found: 85)
```

Similarly, please also rerun `make -C install/kubernetes` and commit the changes (mending the current commit). 

(Edit) You should also run `Documentation/update-spelling_wordlist.sh terminationGracePeriodSeconds` and commit the changes.

**Comment by acgs771126** [comment]
@giorio94 seems resolved all CI failed, please help review the PR again, many thanks.

**Comment by giorio94** [comment]
> @giorio94 seems resolved all CI failure, please help review the PR again, many thanks.

One of the checks is still failing because the commit title is too long: https://github.com/cilium/cilium/actions/runs/5641947820/job/15282283916?pr=26945

**Comment by acgs771126** [comment]
> One of the checks is still failing because the commit title is too long: https://github.com/cilium/cilium/actions/runs/5641947820/job/15282283916?pr=26945

updated the PR, but got Travis-CI failed... 
seems like in [ipam testing step](https://app.travis-ci.com/github/cilium/cilium/builds/264719998#L728), but I don't update any about ipam changes, so maybe you can help re-run the CI testing? or any suggestions.

**Comment by giorio94** [comment]
@acgs771126 Tests are not happy yet :-(. Could you please rebase your branch onto main?

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 3]
```diff
@@ -2718,6 +2718,9 @@ clustermesh:
       # -- Security context to be added to clustermesh-apiserver etcd containers
       securityContext: {}
 
+      # -- lifecycle setting for the etcd container.
```
Comment (zacharysarah) [inline install/kubernetes/cilium/values.yaml.tmpl]: No periods for sentence fragments
```suggestion
      # -- lifecycle setting for the etcd container
```

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 4]
```diff
@@ -2768,6 +2771,9 @@ clustermesh:
           drop:
             - ALL
 
+      # -- lifecycle setting for the KVStoreMesh container.
```
Comment (zacharysarah) [inline install/kubernetes/cilium/values.yaml.tmpl]: No periods for sentence fragments
```suggestion
      # -- lifecycle setting for the KVStoreMesh container
```

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 5]
```diff
@@ -2798,6 +2804,12 @@ clustermesh:
     # -- Number of replicas run for the clustermesh-apiserver deployment.
     replicas: 1
 
+    # -- lifecycle setting for the apiserver container.
+    lifecycle: {}
+
+    # -- terminationGracePeriodSeconds for the clustermesh-apiserver deployment.
```
Comment (zacharysarah) [inline install/kubernetes/cilium/values.yaml.tmpl]: No periods for sentence fragments
```suggestion
    # -- lifecycle setting for the apiserver container
    lifecycle: {}

    # -- terminationGracePeriodSeconds for the clustermesh-apiserver deployment
```

**Review by zacharysarah — CHANGES_REQUESTED** [review]
@acgs771126 Some grammar nits, otherwise LGTM from a docs perspective.

**Review by giorio94 — APPROVED** [review]

**Review by zacharysarah — APPROVED** [review]
LGTM

**Comment by acgs771126** [comment]
@kaworu Hi Alexandre
Sorry to trouble you, hope you have a good vacation.
If your come back to work please help review the PR, thank you :)

File: Documentation/spelling_wordlist.txt [hunk 6]
```diff
@@ -635,6 +635,7 @@ tcptop
 tcptracer
 templating
 terminatingTLS
+terminationGracePeriodSeconds
```
Comment (tommyp1ckles) [inline Documentation/spelling_wordlist.txt:638]: why do we need this, this isnt the first instance of this word

Comment (tommyp1ckles) [reply Documentation/spelling_wordlist.txt:638]: oh nevermind - missed the discussion regarding the commit message ci failure 😅

**Review by tommyp1ckles — APPROVED** [review]
looks good

**Comment by giorio94** [comment]
Reviews are in, and CI is green. Marking as ready to merge.

**Pull request merged**